internal/app/networkd/pkg/vip/vip.go

```diff
@@ -0,0 +1,136 @@
+	"time"
+
+	"github.com/plunder-app/kube-vip/pkg/vip"
+	"github.com/talos-systems/go-retry/retry"
+	"go.etcd.io/etcd/clientv3/concurrency"
+
+	"github.com/talos-systems/talos/internal/pkg/etcd"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+)
+
+const campaignRetryInterval = time.Second
+
+// A Controller provides a control interface for Virtual IP addressing.
+type Controller interface {
+
+	// Start activates the Virtual IP address controller.
+	Start(ctx context.Context) error
+
+	// Stop shuts down the Virtual IP address controller.
+	Stop() error
+}
+
+type vipController struct {
+	ip    net.IP
+	iface *net.Interface
+}
+
+// New creates a new Virtual IP controller.
+func New(ip, iface string) (Controller, error) {
+	ipaddr := net.ParseIP(ip)
+	if ipaddr == nil {
+		return nil, fmt.Errorf("failed to parse ip %q as an IP address", ip)
+	}
+
+	netIf, err := net.InterfaceByName(iface)
+	if err != nil || netIf == nil {
+		return nil, fmt.Errorf("failed to find interface %s by name: %w", iface, err)
+	}
+
+	return &vipController{
+		ip:    ipaddr,
+		iface: netIf,
+	}, nil
+}
+
+// Start implements the Controller interface.
+func (c *vipController) Start(ctx context.Context) error {
+	netController, err := vip.NewConfig(c.ip.String(), c.iface.Name, false)
+	if err != nil {
+		return err
+	}
+
+	go c.maintain(ctx, netController)
+
+	return nil
+}
+
+// Stop implements the Controller interface.
+func (c *vipController) Stop() error {
+	return fmt.Errorf("unimplemented")
+}
+
+func (c *vipController) etcdElectionKey() string {
+	return fmt.Sprintf("%s:vip:election:%s", constants.EtcdRootTalosKey, c.ip.String())
+}
+
+func (c *vipController) maintain(ctx context.Context, netController vip.Network) {
+	for ctx.Err() == nil {
+		if err := c.campaign(ctx, netController); err != nil {
+			log.Println("campaign failure: %w", err)
+
+			time.Sleep(campaignRetryInterval)
+
+			continue
+		}
+	}
+}
+
+func (c *vipController) campaign(ctx context.Context, netController vip.Network) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("refusing to join election without a hostname")
+	}
+
+	var ec *etcd.Client
+
+	err = retry.Constant(5*time.Minute, retry.WithUnits(500*time.Millisecond)).Retry(func() error {
+		ec, err = etcd.NewLocalClient()
+		if err != nil {
+			return retry.ExpectedError(err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to create local etcd client: %w", err)
+	}
+	defer ec.Close() // nolint: errcheck
+
+	sess, err := concurrency.NewSession(ec.Client)
+	if err != nil {
+		return fmt.Errorf("failed to create concurrency session: %w", err)
+	}
+	defer sess.Close() // nolint: errcheck
+
+	election := concurrency.NewElection(sess, c.etcdElectionKey())
+
+	if err := election.Campaign(ctx, hostname); err != nil {
+		return fmt.Errorf("failed to conduct campaign: %w", err)
+	}
+	defer election.Resign(ctx) // nolint: errcheck
+
+	if err := netController.AddIP(); err != nil {
```

Comment:
looks like we're not using any advanced features of kube-vip package here? it's just equivalent to adding/removing an address? unless I'm missing it, looks like `kube-vip` can send ARP messages, etc.?

Reply:
Yeah, the only thing we are really using from `kube-vip` here is the ARP flooding mechanics.  We could, of course, implement the same concept ourselves, but I didn't see a need to reinvent the wheel, off-hand.

Reply:
I don't see it being used here, seems just `AddIP` and `DeleteIP`, unless I'm missing something.

Reply:
Oops.  I don't have the gratuitous ARP in there...